### ISSUE TITLE
Type-hint-driven mismatches are only working for a subset of types

### ISSUE THREAD

**Issue by aisamu** [clj-kondo/clj-kondo#2109]
(opening issue per Slack instructions) 

**version**
`:clj-kondo-version "2023.04.15-SNAPSHOT"`

**platform**
clojure-lsp
`:server-version "2023.05.04-19.38.01"`

**editor**
Emacs

**problem**

Type-hint-driven type mismatches work with `^String`, `^java.lang.String`, `^Double` but not with (e.g.) `java.util.List`

**repro**

```clojure
(defn foo-string [^String s] s)
(foo-string 1)
;; Expected: string or nil

(defn foo-qualified-string [^java.lang.String s] s)
(foo-qualified-string 1)
;; Expected: string or nil 

(defn foo-int [^Long i] (str i))
(foo-int "1")
;; Expected: integer or or nil
;; 
(defn foo-qualified-int [^java.lang.Long i] (str i))
(foo-qualified-int "1")
;; Expected: integer or or nil
 
(foo-int (foo-string 1))
;; Expected: integer or or nil
;; Expected: string or nil 

(def ll (java.util.List/of "1" "2"))
;; => ["1" "2"]
;; i.e. it's on the classpath
 
(defn foo-list [^java.util.List l] l)

(foo-list 1)
(foo-list "1")
;; No complaints
```

```shell
$> clj-kondo --lint src/repro.clj  
WARNING: error while reading /.../.../.clj-kondo/config.edn (EOF while reading)
src/repro.clj:4:13: error: Expected: string or nil, received: positive integer.
src/repro.clj:8:23: error: Expected: string or nil, received: positive integer.
src/repro.clj:12:10: error: Expected: integer or nil, received: string.
src/repro.clj:16:20: error: Expected: integer or nil, received: string.
src/repro.clj:19:10: error: Expected: integer or nil, received: string or nil.
src/repro.clj:19:22: error: Expected: string or nil, received: positive integer.
linting took 14ms, errors: 6, warnings: 0
```

**config**

(no config.edn file)

**expected behavior**

`Expected: java.util.List or nil, received: ______`

Thanks for clj-kondo!
